module/apmchi/middleware.go

```diff
@@ -0,0 +1,104 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmchi
+
+import (
+	"net/http"
+
+	"github.com/go-chi/chi"
+
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmhttp"
+)
+
+// Middleware returns a new chi middleware handler
+// for tracing requests and reporting errors.
+//
+// By default, the middleware will use apm.DefaultTracer.
+// Use WithTracer to specify an alternative tracer.
+func Middleware(o ...Option) func(http.Handler) http.Handler {
+	opts := options{
+		tracer:         apm.DefaultTracer,
+		requestIgnorer: apmhttp.DefaultServerRequestIgnorer(),
+	}
+	for _, o := range o {
+		o(&opts)
+	}
+	return func(h http.Handler) http.Handler {
+		return apmhttp.Wrap(
+			h,
+			apmhttp.WithTracer(opts.tracer),
+			apmhttp.WithServerRequestName(routeRequestName),
+			apmhttp.WithServerRequestIgnorer(opts.requestIgnorer),
+		)
+	}
+}
+
+func routeRequestName(r *http.Request) string {
+	return r.Method + " " + getRequestPattern(r)
+}
+
+func getRequestPattern(r *http.Request) string {
+	rctx := chi.RouteContext(r.Context())
+	if pattern := rctx.RoutePattern(); pattern != "" {
+		return pattern
+	}
+
+	routePath := r.URL.Path
```

Comment:
I'm unfamiliar with Chi, but it appears that if you use `Router.Use`, then the middleware will be run before routing; and if you use `Router.With`, then middleware will run after. Are there any other circumstances in which the route pattern not set above? It doesn't seem ideal to run the route pattern matching twice.

According to https://godoc.org/github.com/go-chi/chi#Context.RoutePattern, the route pattern should be queried after executing the handler. Should the transaction name be set initially to `""` (or perhaps `"\<METHOD\> unknown route"`...), and then updated before ending the transaction?

Another thing to consider is that we may later want to add sampling based on transaction names. The earlier we know the transaction name, the better; that may mean restricting (or recommending) people to add the middleware inline, using `Router.With`.

Reply:
> it appears that if you use Router.Use, then the middleware will be run before routing; and if you use Router.With, then middleware will run after. 

I believe that's true with some caveats. Because of how subroutes are set up, using either `Router.Use` and `Router.With` will have routed to a particular point, but not necessarily the whole thing:

```go
middleware := // ...
r := chi.NewRouter()
r.Use(middleware) // RoutePattern(): ""

r.With(middleware). // RoutePattern(): "/prefix/*"
  Route("/prefix", func(r chi.Router) {
    r.Use(middleware) // RoutePattern(): "/prefix/*"
    r.With(middleware).Get("/{thing}", handler)  // RoutePattern(): "/prefix/{thing}"
  })
```

What should be happening here is:
- If you do `Use` at the top level, there's no route metadata, so it'll find that with `Match`. 
- If you do `With` or do `Use` further down into the routing, it will capture the part of the route pattern so far, but not attempt to find the rest. So you could use it on each route to get the full thing, or on a subrouter if all you care about is the prefix.

I think that's a relatively sane behavior, but I'm not sure if it's desirable/consistent. Other options I can think of are to always match on the full pattern, or start with `unknown route` and swap in the actual route after, like you mentioned.

Either way, let me know, and I'll get some more proper test cases for these situations and update the documentation.

> It doesn't seem ideal to run the route pattern matching twice.

Agreed, but that's [the recommended pattern](https://github.com/go-chi/chi/issues/270#issuecomment-335570862) as far as I can tell, and it's [used in the official middleware](https://github.com/go-chi/chi/blob/master/middleware/get_head.go#L29) for similar purposes.

> Should the transaction name be set initially to "" (or perhaps "\<METHOD\> unknown route"...), and then updated before ending the transaction?

I'm not sure exactly what the trade-offs here are from an APM perspective, but if this is the preferred behavior, it should be as easy as re-implementing a slightly modified version of the `handler.ServeHTTP` function from `apmhttp`. Of course, that would mean not knowing the transaction name until the transaction is basically over.

So I guess there's options—let me know what you think is the way forward.

Reply:
> Agreed, but that's the recommended pattern as far as I can tell, and it's used in the official middleware for similar purposes.

OK, fair enough. Thanks for the detailed reply.

The difference in behaviour between `Use` and `With` seems a little subtle to me; generally transactions are named after the full matched route. I would just go with always matching the full route pattern, but if you have a use case for naming transactions by a route prefix, then I'm fine with keeping that flexibility.

> Either way, let me know, and I'll get some more proper test cases for these situations and update the documentation.

Thank you, that would be excellent.

Reply:
> I would just go with always matching the full route pattern

Yeah I think this is the behavior that makes the most sense. Done.